standup/auth0/middleware.py

```diff
@@ -44,78 +45,84 @@ def renew_id_token(id_token):
-    parsed = urlparse(url)
-    if parsed.query:
-        return '%s?%s' % (parsed.path, parsed.query)
-    return parsed.path
+class ValidateIdToken(object):
+    """For users authenticated with an id_token, we need to check that it's still valid. For
+    example, the user could have been blocked (e.g. leaving the company) if so we need to ask the
+    user to log in again.
 
+    We do this using cache.
 
-class ValidateIDToken(object):
-    """For users authenticated with an id_token, we need to check that it's still valid. For example,
-    the user could have been blocked (e.g. leaving the company) if so we need to ask the user to log
-    in again.
+    # FIXME(willkg): Rework this so that the cache parts are methods that can be overridden.
 
     """
 
     exception_paths = (
-        get_path(settings.AUTH0_CALLBACK),
+        # Exclude the AUTH0_CALLBACK_URL path, otherwise this can loop
+        urlparse(settings.AUTH0_CALLBACK_URL).path,
     )
 
     def process_request(self, request):
         if (
-            request.method != 'POST' and
-            not request.is_ajax() and
-            request.user.is_active and
-            request.path not in self.exception_paths
+                request.method != 'POST' and
+                # FIXME(willkg): We might want to do this for AJAX, too, otherwise one-page webapps
+                # might never renew.
+                not request.is_ajax() and
+                request.user.is_active and
+                request.path not in self.exception_paths
         ):
-            # Look up expiration in session and see if the id_token needs to be renewed.
-            id_token_expiration = request.session.get('id_token_expiration', None)
-            if id_token_expiration and id_token_expiration < datetime.utcnow():
+            cache_key = 'auth0:renew_id_token:%s' % request.user.id
+
+            # Look up expiration in cache to see if id_token needs to be renewed.
+            if cache.get(cache_key):
                 return
 
-            # Either no expiration in session or token needs to be renewed, so renew it
-            # now.
+            # The id_token has expired, so we renew it now.
             try:
                 token = IdToken.objects.get(user=request.user)
             except IdToken.DoesNotExist:
-                # If there is no IdToken, then this isn't a mozilla.com address and we're fine.
+                # If there is no IdToken, check if this domain requires them and if so, redirect and
+                # if not, we're fine.
+                if request.user.email:
+                    domain = request.user.email.lower().split('@', 1)[1]
+                    if domain in settings.AUTH0_ID_TOKEN_DOMAINS:
+                        messages.error(
+                            request,
+                            'You can\'t log in with that email address using the provider you '
+                            'used. Please log in with the Oauth2 provider.'
```

Comment:
This message doesn't really work for Mozilla because it doesn't match the auth0 interface users see, so we'll need to figure out what to do about that.